src/types/profile.js

```diff
@@ -15,17 +15,56 @@ export type IndexIntoStringTable = number;
 export type IndexIntoFuncTable = number;
 export type IndexIntoResourceTable = number;
 export type IndexIntoLibs = number;
-export type categoryBitMask = number;
+export type IndexIntoCategoryList = number;
 export type resourceTypeEnum = number;
 export type ThreadIndex = number;
 
 /**
- * The stack table is the minimal representation of a call stack. Each stack entry
- * consists of the frame at the top of the stack, and the prefix for the stack that
- * came before it. Stacks can be shared between samples.
+ * The stack table stores the tree of stack nodes of a thread.
+ * The shape of the tree is encoded in the prefix column: Root stack nodes have
+ * null as their prefix, and every non-root stack has the stack index of its
+ * "caller" / "parent" as its prefix.
+ * Every stack node also has a frame and a category.
+ * A "call stack" is a list of frames. Every stack index in the stack table
+ * represents such a call stack; the "list of frames" is obtained by walking
+ * the path in the tree from the root to the given stack node.
+ *
+ * Stacks are used in the thread's samples; each sample refers to a stack index.
+ * Stacks can be shared between samples.
+ *
+ * With this representation, every sample only needs to store a single integer
+ * to identify the sample's stack.
+ * We take advantage of the fact that many call stacks in the profile have a
+ * shared prefix; storing these stacks as a tree saves a lot of space compared
+ * to storing them as actual lists of frames.
+ *
+ * The category of a stack node is always non-null and is derived from a stack's
+ * frame and its prefix. Frames can have null categories, stacks cannot. If a
+ * stack's frame has a null category, the stack inherits the category of its
+ * prefix stack. Root stacks whose frame has a null stack have their category
+ * set to the "default category". (The default category is currently defined as
+ * the category in the profile's category list whose color is "grey", and such
+ * a category is required to be present.)
+ *
+ * You could argue that the stack table's category column is derived data and as
+ * such doesn't need to be stored in the profile itself. This is true, but
+ * storing this information in the stack table makes it a lot easier to carry
+ * it through various transforms that we apply to threads.
+ * For example, here's a case where a stack's category is not recoverable from
+ * any other information in the transformed thread:
+ * In partial the call path
```

Comment:
nit: I don't understand this "In partial". Should this be something like `When parsing` ?

Reply:
Oops, this was supposed to say "In *the* partial call path [...]". I'm going to change it back to "In the call path [...]" which is what I had there first.
(I added the word "partial" because that path, as written, doesn't actually occur in real Gecko profiles because I left out a few functions in between.)